_posts/2021-04-29-gpu-glossary.md

```diff
@@ -0,0 +1,197 @@
+* [WebGPU](https://github.com/gpuweb/gpuweb) - [WebGPU shading language](https://gpuweb.github.io/gpuweb/wgsl/)
+* [CUDA](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html)
+
+My choice of 5 API's is somewhat opinionated, and future-leaning. I've left off OpenGL and OpenCL because they appear to be stagnating. Even though my main focus is writing compute shaders, because I'm interested in portability, I'm also including CUDA because it's by far the best developed API for writing GPU compute, and one that will no doubt be familiar to mamy readers.
+
+I'm including WebGPU because I feel it has potential to become a common, widely implemented standard, on which it's possible to build a rich infrastructure of tools, libraries, and applications. The people working on WebGPU are doing a lot of deep thinking about achieving portability, performance, and security on GPU, and these discussions are openly available.
+
+Another resource I've found invaluable for writing (somewhat) portable compute shader code is [gfx-hal]. This project exposes an API much like Vulkan (though not identical), with [backends](https://github.com/gfx-rs/gfx/tree/master/src/backend) for all the major API's. In many cases, the best way to answer the question "how to I do X in API Y" is to find the gfx-hal method for it and look up the code in the respective backend.
+
+In general I favor Vulkan terminology, as it's an open standard, well documented, and in some areas (particularly memory model) is significantly ahead of its competition.
+
+This post also especially welcomes open source contributions, as no doubt some of the entries are incomplete or confusing. File an issue or a PR at https://github.com/raphlinus/raphlinus.github.io.
+
+## The thread grouping hierarchy
+
+A defining characteristic of GPU compute is a fairly deep hierarchy of groups of threads. At each level, there is interaction with threads in the same group, very high bandwidth and low latency at the bottom of the hierarchy, access to more of the workload as you go further up. But the programming interfaces (and terminology) varies, reflecting the realities of GPU hardware.
+
+We'll take this hierarchy from the bottom up.
+
+### Thread; Invocation (Vulkan, formal); Lane (DX)
+
+A thread in GPU compute is *like* a traditional CPU thread, but with some differences. While on CPU, the cost of spawning a thread is significant, on a GPU it is quite cheap, and often the "many small threads" pattern is the best way to achieve performance.
+
+GPU threads are scheduled quite differently than on CPU. For one, forward progress guarantees are much weaker (though this depends on the GPU). The hardware model of GPU is often described as [Single Instruction Multiple Threads] (SIMT), which is similar to more traditional Single Instruction Multiple Data (SIMD), but with more support for creating the illusion that the threads are running independently, in particular execution masks for conditional control flow.
```

Comment:
I'm having trouble reading the end of the last sentence (is there a verb missing?)... but it might just be me